src/interface/params.hpp

```diff
@@ -108,6 +112,40 @@ class Params {
     std::cout << std::endl;
   }
 
+#ifdef HDF5OUTPUT
+
+  // will write all params with type T to the given HDF5 group as an attribute
+  template <typename T>
+  void WriteToHDF5(const std::string &prefix, const HDF5::H5G &group) const {
+    for (const auto &p : myParams_) {
+      const auto &key = p.first;
+      const auto type = myTypes_.at(key);
+      if (type == std::type_index(typeid(T))) {
+        auto typed_ptr = dynamic_cast<Params::object_t<T> *>((p.second).get());
+        HDF5::HDF5WriteAttribute(prefix + "/" + key, *typed_ptr->pValue, group);
+      }
+    }
+  }
+
+  template <typename T>
+  void WriteTVecToHDF5(const std::string &prefix, const HDF5::H5G &group) const {
+    WriteToHDF5<T>(prefix, group);
+    WriteToHDF5<std::vector<T>>(prefix, group);
+  }
+
+  void WriteAllToHDF5(const std::string &prefix, const HDF5::H5G &group) const {
+    WriteTVecToHDF5<bool>(prefix, group);
+    WriteTVecToHDF5<int32_t>(prefix, group);
+    WriteTVecToHDF5<int64_t>(prefix, group);
+    WriteTVecToHDF5<uint32_t>(prefix, group);
+    WriteTVecToHDF5<uint64_t>(prefix, group);
+    WriteTVecToHDF5<float>(prefix, group);
+    WriteTVecToHDF5<double>(prefix, group);
+    WriteTVecToHDF5<std::string>(prefix, group);
+  }
```

Comment:
That's a good point. I think I should rename `WriteToHDF5` and `WriteTVecToHDF5`, because it seems confusing. These functions don't write specific parameters you specify, instead, they write all parameters that match the type you give it. So if you ask it to write all, say, `int`'s and there are no `int`'s, that's fine. I don't think there should be a warning for that.

Also, I'm thinking of `WriteToHDF5` and `WriteTVecToHDF5` as purely internal functions that should only be used by `WriteAllToHDF5`. The reason they are separate functions is to reduce code duplication.

Let me see if I can give them a better name and encapsulate them better, so they won't be available where they shouldn't be used.

Reply:
> Also, I'm thinking of `WriteToHDF5` and `WriteTVecToHDF5` as purely internal functions that should only be used by `WriteAllToHDF5`. The reason they are separate functions is to reduce code duplication.

I realized that later in the review - I think you can just make those two `private`, that would be fine

Reply:
I've made them `private` and given them a much more descriptive name